File: puzzles/medium/js/skynet-revolution1.js
class Link {
  constructor(node1, node2) {
    this.node1 = node1;
    this.node2 = node2;
  }

  // sever the link between two nodes
  slice() {
    print(this.node1 + ' ' + this.node2);
  }

  equals(link) {
    return ((this.node1 === link.node1) && (this.node2 === link.node2)) ||
           (((this.node1 === link.node2) && (this.node2 === link.node1)));
  }
}

let graph = [];
let gateways = [];

const inputs = readline().split(' ');
const nbNodes = parseInt(inputs[0]); // the total number of nodes in the level, including the gateways
const nbLinks = parseInt(inputs[1]); // the number of links
const nbExits = parseInt(inputs[2]); // the number of exit gateways

for (let i = 0; i < nbLinks; ++i) {
  let inputs = readline().split(' ');
  let N1 = parseInt(inputs[0]); // N1 and N2 defines a link between these nodes
  let N2 = parseInt(inputs[1]);
  graph.push(new Link(N1, N2));
  printErr(graph[i].node1 + ' ' + graph[i].node2);
}

for (let i = 0; i < nbExits; ++i) {
  let gateway = parseInt(readline()); // the index of a gateway node
  gateways.push(gateway);
  printErr(gateways[i]);
}

// game loop
while (true) {
  let agentNode = parseInt(readline()); // node index on which the Skynet agent is positioned this turn
  blockAgent(agentNode);
}

function blockAgent(agentNode) {
  let link = new Link();
  for (let i = 0; i < gateways.length; ++i) {
    let gateway = gateways[i];
    link.node1 = agentNode;
    link.node2 = gateway;
    // Check if there is a link between the agent and the gateway nodes
    for (let j = 0; j < graph.length; ++j) {
      let current = graph[j];
      if (link.equals(current)) {
        link.slice();
        graph.splice(j, 1);
        return;
      } // if
    } // for
  } // for
  blockFirstLink();
} // blockAgent()

function blockFirstLink() {
  graph[0].slice();
  graph.splice(0, 1);
}

File: puzzles/cpp/skynet-revolution1.cpp
#include <iostream>
#include <vector>
#include <utility>

using namespace std;

void sever(pair<int, int>& link) {
  cout << link.first << ' ' << link.second << endl;
}

void block_first_link(vector<pair<int, int>>& links) {
  sever(links[0]);
  links.erase(links.begin());
}

void block(int agent_node, vector<int>& gateways, vector<pair<int, int>>& links) {
  for (auto gateway : gateways) {
    // Check for a direct link between the agent and a gateway
    for (unsigned i = 0; i < links.size(); i++) {
      pair<int, int>& link = links[i];
      if (link == make_pair(agent_node, gateway) || link == make_pair(gateway, agent_node)) {
        sever(link);
        links.erase(links.begin() + i);
        return;
      }
    }
  }

  block_first_link(links);
}

int main() {
  vector<pair<int, int>> links;
  vector<int> gateways;

  int nb_nodes;
  int nb_links;
  int nb_exits;
  cin >> nb_nodes >> nb_links >> nb_exits; cin.ignore();
  for (int i = 0; i < nb_links; i++) {
    int n1;
    int n2;
    cin >> n1 >> n2; cin.ignore();
    pair<int, int> link = make_pair(n1, n2);
    links.push_back(link);
  }

  for (int i = 0; i < nb_exits; i++) {
    int gateway; // the index of a gateway node
    cin >> gateway; cin.ignore();
    gateways.push_back(gateway);
  }

  // game loop
  while (true) {
    int agent_node; // The index of the node on which the Skynet agent is positioned this turn
    cin >> agent_node; cin.ignore();
    block(agent_node, gateways, links);
  }
}
